jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java

```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
         } else if (l.equals("unsat")) {
           res = Result.UNSAT;
           if (unsatCoreSolver) {
+            unsatCoreLastRun = new ArrayList<>();
+            unsatCoreLastRun.add(currentRun);
             ctx.getUnsatCore();
-            String core = "";
-            while (!solverOutput.ready() && p.isAlive()) {
-              Thread.sleep(10);
-            }
-            while (solverOutput.ready() && (line = solverOutput.readLine()) != null) {
-              core += line;
-              core += "\n";
-            }
-            analyzeCore(core);
+            analyzeCore(readProcessOutput(ctx, solverOutput));
           }
         }
       }
-      return res;
-    } catch (IOException | InterruptedException e) {
-      e.printStackTrace();
     }
-    return Result.ERROR;
+    return res;
+  }
+
+  protected static String readProcessOutput(
+      SMTLibExportGenContext context, BufferedReader processOut)
+      throws ExecutionException, InterruptedException, TimeoutException {
+    StringBuilder output = new StringBuilder();
+    String line;
+    String uid = "a" + UUID.randomUUID();
+    boolean matchedUid = false;
+    ExecutorService threadExecutor = Executors.newCachedThreadPool();
```

Comment:
Each invocation appears to create a Cache + Thread, which is immediately discarded afterwards. Why not a field to actually use the cache?